Change GCE to use standalone Saltstack config

zmerlynn opened the pull request:
Change provisioning to pass all variables to both master and node. Run Salt in a masterless setup on all nodes ala http://docs.saltstack.com/en/latest/topics/tutorials/quickstart.html, which involves ensuring Salt daemon is NOT running after install. Kill Salt master install. And fix push to actually work in this new flow.

As part of this, the GCE Salt config no longer has access to the Salt mine, which is primarily obnoxious for two reasons:
- The minions can't use Salt to see the master: this is easily fixed by static config.
- The master can't see the list of all the minions: this is fixed temporarily by static config in util.sh, but later, by other means (see pieces of #156, which should eventually remove this direction).

Along the way: refactor the node setup/push script(s) in util.sh.

Fixes #5112.

zmerlynn committed:
Change GCE to use standalone Saltstack config:

Change provisioning to pass all variables to both master and node. Run
Salt in a masterless setup on all nodes ala
http://docs.saltstack.com/en/latest/topics/tutorials/quickstart.html,
which involves ensuring Salt daemon is NOT running after install. Kill
Salt master install. And fix push to actually work in this new flow.

As part of this, the GCE Salt config no longer has access to the Salt
mine, which is primarily obnoxious for two reasons: - The minions
can't use Salt to see the master: this is easily fixed by static
config. - The master can't see the list of all the minions: this is
fixed temporarily by static config in util.sh, but later, by other
means (see
https://github.com/GoogleCloudPlatform/kubernetes/issues/156, which
should eventually remove this direction).

As part of it, flatten all of cluster/gce/templates/* into
configure-vm.sh, using a single, separate piece of YAML to drive the
environment variables, rather than constantly rewriting the startup
script.

zmerlynn commented:
I have no idea who's a good candidate to review this. Perhaps @filbranden for bash (though a lot of it's just shuffled around), or @erictune to tell me this is a terrible approach.

filbranden commented on cluster/gce/templates/common.sh:
```diff
@@ -44,17 +42,26 @@ install-salt() {
+  echo manual > /etc/init/salt-minion.override
+
+  service salt-minion stop
+  until ! service salt-minion status >/dev/null; do
```
Why not `while` instead of `until !` ?

zmerlynn replied:
I think I had `until` on the brain from the curl loops in the startup script. :)

filbranden commented on cluster/gce/templates/create-salt-files.sh:
```diff
@@ -33,21 +35,19 @@ enable_cluster_dns: '$(echo "$ENABLE_CLUSTER_DNS" | sed -e "s/'/''/g")'
-mkdir -p /srv/salt-overlay/salt/nginx
-echo $MASTER_HTPASSWD > /srv/salt-overlay/salt/nginx/htpasswd
+# This should only happen on cluster initialization
+if [[ ! -z "${KUBELET_TOKEN:-}" ]]; then
```
Use `[[ -n ... ]]` instead of `! -z`

zmerlynn replied:
Done

filbranden commented on cluster/gce/templates/create-salt-files.sh:
```diff
@@ -33,21 +35,19 @@ enable_cluster_dns: '$(echo "$ENABLE_CLUSTER_DNS" | sed -e "s/'/''/g")'
+# This should only happen on cluster initialization
+if [[ ! -z "${KUBELET_TOKEN:-}" ]]; then
+  mkdir -p /srv/salt-overlay/salt/nginx
+  echo $MASTER_HTPASSWD > /srv/salt-overlay/salt/nginx/htpasswd
```
Double quotes around the variable.

zmerlynn replied:
Done

filbranden commented on cluster/gce/templates/create-salt-files.sh:
```diff
@@ -33,21 +35,19 @@ enable_cluster_dns: '$(echo "$ENABLE_CLUSTER_DNS" | sed -e "s/'/''/g")'
-kubelet_token=$(cat /dev/urandom | base64 | tr -d "=+/" | dd bs=32 count=1 2> /dev/null)
+  mkdir -p /srv/salt-overlay/salt/kube-apiserver
+  known_tokens_file="/srv/salt-overlay/salt/kube-apiserver/known_tokens.csv"
+  (umask u=rw,go= ; echo "${KUBELET_TOKEN},kubelet,kubelet" > $known_tokens_file)
```
I think `umask 077` is more clear in its intent. In ~15 years with Linux this is the first time I see this kind of usage...

Also, double quotes around the `${known_tokens_file}` variable.

zmerlynn replied:
Done

filbranden commented on cluster/gce/templates/create-salt-files.sh:
```diff
@@ -33,21 +35,19 @@ enable_cluster_dns: '$(echo "$ENABLE_CLUSTER_DNS" | sed -e "s/'/''/g")'
-(umask u=rw,go= ; echo "{\"BearerToken\": \"$kubelet_token\", \"Insecure\": true }" > $kubelet_auth_file)
+  mkdir -p /srv/salt-overlay/salt/kubelet
+  kubelet_auth_file="/srv/salt-overlay/salt/kubelet/kubernetes_auth"
+  (umask u=rw,go= ; echo "{\"BearerToken\": \"${KUBELET_TOKEN}\", \"Insecure\": true }" > $kubelet_auth_file)
```
Same as above.

zmerlynn replied:
Done

filbranden commented:
No huge red flags from me, except, well, the whole idea of using shell scripts to generate shell scripts... But that's pretty much pre-existing, so probably out of scope for this PR. I made a note of a few nitpicks that could be addressed if you'd like.

derekwaynecarr commented:
Is this a more reliable setup in general?

I have had troubles in past where a salt minion randomly flakes when communicating to master.

Are updates any faster?

If so, would do this with Vagrant when I have time.

zmerlynn commented:
@derekwaynecarr: I didn't measure the `kube-push` flow prior to making the update, if that's what you're asking about. The bring-up flow is pretty similarly timed. As for the actual reliability, it's hard to measure yet, but I'd imagine it should be. The `salt-call --local` should be deterministic with the variables set the way they are. I was also somewhat aggressive in this PR and changed our startup script to `errexit` prior to the salt call so we'd get an early warning if something else was wrong, but did hesitate to do that for the salt call itself. I may go back after this and futz with the return value off Salt as well, but the PR was creeping enough. (I think Salt tends to claim errors for things that aren't actually "in error" in my mind, though, like daemons that just haven't been started yet, because it's trying to reach desired state. So a second call would actually be a reliable exit code.)

zmerlynn commented:
@filbranden: Yeah, I considered moving the startup script to be a static entity that gets uploaded and just have the env variables be poked into the metadata server, but you actually really don't want to have to poke individual things into each instance on the metadata server. We're trying to move away from that for the CIDR selection soon, then MIG will just have the one static startup script (that was generated by bash). But I hear you, the bash-generating-bash blew my mind the first couple of times. Actually, that says it's at least worth a comment in each of the composed scripts that says "BTW, this script is transcluded in the startup/push script".

Addressing your other comments now.

filbranden commented:
@zmerlynn Yeah ideally data should be pushed as a single metadata value as either JSON or YAML, and the script/binary on the other side should use a language decent enough to parse it (which excludes bash and solves a lot of the problem!)

Though this is mostly pre-existing and out of scope for this PR.

Thanks for addressing my comments.

As I mentioned, bash script LGTM, just not sure who is the best reviewer for the actual change...

Cheers,
Filipe

zmerlynn commented:
Looks like @brendanburns stepped up. :)

zmerlynn commented:
Hm. It's definitely beyond this PR, but it looks like you can actually set common metadata on the the MIG (which is an obvious use-case). So it would be possible to clean this up to a static startup script that derived its variables a reasonable way, and I have an idea how. Baby steps.

zmerlynn commented:
@filbranden / @brendanburns: Actually, I pounded out a YAML metadata version on Friday afternoon / this morning. There's a commit on top of the last commit that I sent. I can either split it out of this PR or we can do it here. (I also squashed the review comments into the first commit, since you LGTM'd after.)

This actually makes my life a lot simpler on the GKE side, since we'd then be deploying using, *gasp*, as you suggested, a single metadata value (except for the weird bit about the kubelet token). It's still bash receiving it, but that's not the end of the world except for the nuance of it being all strings, all the the time.

brendandburns commented on cluster/gce/configure-node.sh:
```diff
@@ -0,0 +1,352 @@
+set -o pipefail
+
+# If we have any arguments at all, this is a push and not just setup.
+is_push=$*
```
```$@```

brendandburns commented on cluster/gce/configure-node.sh:
```diff
@@ -0,0 +1,352 @@
+}
+
+function set-good-motd() {
+  echo "echo -e '\n=== GCE Kubernetes node setup complete ===\n' > /etc/motd"
```
double echo?

brendandburns commented on cluster/gce/configure-node.sh:
```diff
@@ -0,0 +1,352 @@
+}
+
+function curl-metadata() {
+  curl --fail --silent -H 'Metadata-Flavor: Google' "http://metadata/computeMetadata/v1/instance/attributes/$1"
```
${1}

brendandburns commented on cluster/gce/configure-node.sh:
```diff
@@ -0,0 +1,352 @@
+  until [[ -e "${1##*/}" ]]; do
+    echo "Downloading file ($1)"
+    curl --ipv4 -Lo "$file" --connect-timeout 20 --retry 6 --retry-delay 10 "$1"
+    md5sum "$file"
```
if we don't validate this, there's little point in md5summing.

brendandburns commented on cluster/gce/configure-node.sh:
```diff
@@ -0,0 +1,352 @@
+
+  mkdir -p /srv/salt-overlay/salt/kube-apiserver
+  known_tokens_file="/srv/salt-overlay/salt/kube-apiserver/known_tokens.csv"
+  (umask 077; echo "${KUBELET_TOKEN},kubelet,kubelet" > "${known_tokens_file}")
```
line break after semi-colon

brendandburns commented on cluster/gce/configure-node.sh:
```diff
@@ -0,0 +1,352 @@
+
+  mkdir -p /srv/salt-overlay/salt/kubelet
+  kubelet_auth_file="/srv/salt-overlay/salt/kubelet/kubernetes_auth"
+  (umask 077; echo "{\"BearerToken\": \"${KUBELET_TOKEN}\", \"Insecure\": true }" > "${kubelet_auth_file}")
```
line break after semi-colon

brendandburns commented:
LGTM, small stuff.

zmerlynn commented:
PTAL. Running e2es now.

zmerlynn commented:
e2es pass in mixed runs.

zmerlynn committed:
One additional cleanup: Send only the master name, rather the internal
IP address. The configure-vm script can resolve this relatively easily
on the node. This is less painful for GKE, which creates all the
resources in parallel.

zmerlynn commented:
@brendanburns: Realized it trivial just to pass the master name, too, and it cleans things up nicely. It lightens the `util.sh` delta as well. I can either keep that commit here or break it off to another PR. PTAL and merge when you're happy (e2es pass on the combined)
